Use annotation mechanism for suitable plugins

chmeliik opened the pull request:
* OSBS-7828

`[WIP]` commits contain TODOs explaining what is still WIP.

Also, if anyone has a better idea for the `annotation_map` and `label_map` function names, that would be welcome.

Maintainers will complete the following section:
- [x] Commit messages are descriptive enough
- [x] "Signed-off-by:" line is present in each commit
- [x] Code coverage from testing does not decrease and new code is covered
- [x] JSON/YAML configuration changes are updated in the relevant schema
- [x] Changes to metadata also update the documentation for the metadata
- [x] Pull request includes link to an osbs-docs PR for user documentation updates
- [x] New feature can be disabled from a configuration file

chmeliik commented:
It appears that returning `'help_file': None` from `add_help` plugin is indeed OK

chmeliik commented:
I'm assuming labels will always be simple values (numbers, strings) which is why did not `json.dumps` them but simply convert to `str`. Dropping the remaining TODOs.

chmeliik commented:
Rebased on master

twaugh commented:
Nice!

Seeing it in action I wonder whether the decorators should live in a new `metadata.py` module,
```
from atomic_reactor import metadata

@metadata.label(...)
class ...
```

What do you think?

chmeliik commented:
> Looks really nice! Think you'll edit the commits to add things in logical order, or keep them as they are?

Can do, would the logical order be `update decorators -> move to new module -> update plugins`?

twaugh commented:
Or move, update decorators, update plugins.

twaugh commented on atomic_reactor/util.py:
```diff
@@ -1591,33 +1591,47 @@ def get_platform_config(platform, build_annotations):
     >>> class MyBuildPlugin(BuildPlugin):
     >>>     key = 'my_build_plugin'
     >>>
+    >>>     # sets annotations: {'foo': 1, 'bar': 1}
```
Not sure about this. It sets the same value on both annotations? It doesn't seem like useful behaviour to duplicate it.

Should it just take a single argument (the annotation name)? If it takes multiple arguments I'd expect it to unpack a tuple from the return value.

chmeliik replied:
Fair point, I don't think I've seen any plugins returning tuples so I will probably just make it take only a single argument.

chmeliik commented:
Reordered commits, I went with update first, then move (which was easier), hope you don't mind

chmeliik commented:
Made `annotation` and `label` accept only 1 key

MartinBasti commented on atomic_reactor/util.py:
```diff
@@ -1616,29 +1616,53 @@ def annotation(*keys):
     from atomic_reactor.plugin import BuildPlugin
 
-    def annotate(cls):
+    wrong_class = '[{}] Not a subclass of BuildPlugin'.format(metadata_type)
```
I'd prefer lazy formatting here

MartinBasti commented on atomic_reactor/plugins/pre_add_filesystem.py:
```diff
@@ -33,10 +33,12 @@
 from osbs.utils import Labels
 
 
+@label_map('filesystem-koji-task-id')
```
Should we have constants for label and annotations keys?

chmeliik commented:
Added test and changed where error messages are created.

As for constants, this actually reduces the need (because the `store_metadata_in_osv3` plugin no longer needs to know what the annotations and labels are called) but I can add them if they are needed anyway.

chmeliik commented:
Force-pushed for coveralls results

MartinBasti commented:
Internal change, no release notes needed
